src/main/webapp/scripts/script.js

```diff
@@ -0,0 +1,132 @@
+      var email = user.email;
+      var emailVerified = user.emailVerified;
+      var photoURL = user.photoURL;
+      var isAnonymous = user.isAnonymous;
+      var uid = user.uid;
+      var providerData = user.providerData;
+
+      document.getElementById('login').textContent = 'Log out';
+      document.getElementById('login-status').textContent = 'Signed in.';
+    } else {  // User signed out.
+      document.getElementById('login').textContent = 'Log in';
+      document.getElementById('login-status').textContent = 'Signed out.';
+    }
+    document.getElementById('login').disabled = false;
+  });
+}
+
+/** 
+ * If user is currently logged in, log them out; if logged out, log them in.
+ */
+function toggleLogin() {
+  // If the user is signed in, sign them out.
+  if (firebase.auth().currentUser) {
+    firebase.auth().signOut();
+  } else {  // Sign in.
+    var email = document.getElementById('email').value;
+    var password = document.getElementById('password').value;
+
+    if (email == null || email.length < 2) {
+      alert('Please enter an email address.');
+      return;
+    }
+    if (password == null || password.length < 2) {
+      alert('Please enter a password.');
+      return;
+    }
+
+    firebase.auth()
+        .signInWithEmailAndPassword(email, password)
+        .then(function onSuccess() {
+          // TODO: send token to server
+            // firebase.auth()
+            //   .currentUser.getIdToken(/* forceRefresh */ true)
+            //   .then(function(idToken) {
+            //     // Body type defaults to JSON.
+            //     fetch('/login', {method: 'POST', body: idToken});
+            //   })
+            //   .catch(function onFailure(error) {
+            //     alert('Token retrieval failed: ' + error);
+            //   }); 
+        })
+        .catch(function onFailure(error) {
+          var errorCode = error.code;
+          var errorMessage = error.message;
+
+          if (errorCode === 'auth/wrong-passcode') {
+            alert('Wrong password.');
+          } else {
+            alert(errorMessage);
+          }
+        });
+    // document.getElementById("login").disabled = true;
+  }
+}
+
+/**
+ * Attempt to create a new user account using the email/password HTML fields.
+ */
+function onSignup() {
+  var email = document.getElementById("email").value;
+  var password = document.getElementById("password").value;
+
+  if (email == null || email.length < 2) {
+    alert("Please enter a valid email address.");
+  }
+  if (password == null || password.length < 2) {
+    alert("Please enter a valid password.");
+  }
+  
+  // If email/pass are of appropriate length, try to create a new account.
+  firebase.auth().createUserWithEmailAndPassword(email, password)
+      .catch(function onFailure(error) {
+        var errorCode = error.code;
+        var errorMessage = error.message;
+
+        if (errorCode == 'auth/weak-password') {
+          alert('The password is too weak.');
```

Comment:
Why are we using a custom error message only for auth/weak-password? Shouldn't we always use error.message?

Reply:
I'm using a custom error message for the more common errors since I think they're probably more user-friendly than whatever the default is. Some of the errors authentication throws are very niche, like auth/missing-ios-bundle-id, so for those I'm not as worried about a custom message since a user probably shouldn't see it anyway